src/extension/index.activateDecorations.ts

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+/* eslint-disable filenames/match-regex */
+
+import { IArraySplice, observable, observe } from 'mobx';
+import { Log } from 'sarif';
+import { Disposable, languages, Range, ThemeColor, window } from 'vscode';
+import { findResult, parseArtifactLocation, ResultId } from '../shared';
+import '../shared/extension';
+import { regionToSelection } from './regionToSelection';
+import { ResultDiagnostic } from './resultDiagnostic';
+import { Store } from './store';
+
+// Decorations are for Analysis Steps.
+export function activateDecorations(disposables: Disposable[], store: Store) {
+    // Navigating away from a diagnostic/result will not clear the `activeResultId`.
+    // This keeps the decorations "pinned" while users navigate the thread flow steps.
+    const activeResultId = observable.box<string | undefined>();
+
+    const decorationTypeCallout = window.createTextEditorDecorationType({
+        after: { color: new ThemeColor('problemsWarningIcon.foreground') }
+    });
+    const decorationTypeHighlight = window.createTextEditorDecorationType({
+        border: '1px',
+        borderStyle: 'solid',
+        borderColor: new ThemeColor('problemsWarningIcon.foreground'),
+    });
+
+    // On selection change, set the `activeResultId`.
+    disposables.push(languages.registerCodeActionsProvider('*', {
+        provideCodeActions: (_doc, _range, context) => {
+            if (context.only) return;
+
+            const diagnostic = context.diagnostics[0] as ResultDiagnostic | undefined;
+            if (!diagnostic) return;
+
+            const result = diagnostic?.result;
+            if (!result) return; // Don't clear the decorations. See `activeResultId` comments.
+
+            activeResultId.set(JSON.stringify(result._id)); // Stringify for comparability.
+
+            // Technically should be using `onDidChangeTextEditorSelection` and `languages.getDiagnostics`
+            // then manually figuring with diagnostics are at the caret. However `languages.registerCodeActionsProvider`
+            // provides the diagnostics for free. The only odd part is that we always return [] when `provideCodeActions` is called.
+            return [];
+        }
```

Comment:
The code actions provider still causes unexpected switches between results when going through the analysis steps in some cases. (see #441) Wouldn't it make more sense to make the panel webview include the result ID in the `select` messages and set `activeResultId` in the message handler? I'm not experienced with VS Code extensions, so maybe there's something I'm missing.

Reply:
@keremc The root issue is when multiple Results overlap the same range of text we do not support selecting a specific Result within that set. I put more details in #441.

The `select` messages are used for more than just selecting Results. If we are to include additional data, I would prefer to account for all the cases.

Also, `activeResultId` is local to the Decorations feature. `activeResultId` is created out of convenience as an event source. The data it represents is derived, and normally I would prefer not to store this.